### PR TITLE
Improve pushdown optimization and logical to physical transformation

### PR DESCRIPTION
**Pull request by dai-chen**
Signed-off-by: Chen Dai <daichen@amazon.com>

### Background

This section introduces the current architecture of logical optimizer and physical transformation.

#### Logical-to-Logical Optimization

Currently each storage engine adds its own logical operator as concrete implementation for `TableScanOperator` abstraction. Typically each data source needs to add 2 logical operators for table scan with and without aggregation. Take OpenSearch for example, there are `OpenSearchLogicalIndexScan` and `OpenSearchLogicalIndexAgg` and a bunch of pushdown optimization rules for each accordingly.

```
class LogicalPlanOptimizer:
  /*
   * OpenSearch rules include:
   *   MergeFilterAndRelation
   *   MergeAggAndIndexScan
   *   MergeAggAndRelation
   *   MergeSortAndRelation
   *   MergeSortAndIndexScan
   *   MergeSortAndIndexAgg
   *   MergeSortAndIndexScan
   *   MergeLimitAndRelation
   *   MergeLimitAndIndexScan
   *   PushProjectAndRelation
   *   PushProjectAndIndexScan
   *
   * that return *OpenSearchLogicalIndexAgg*
   *  or *OpenSearchLogicalIndexScan* finally
   */
  val rules: List<Rule>

  def optimize(plan: LogicalPlan):
    for rule in rules:
      if rule.match(plan):
        plan = rules.apply(plan)
    return plan.children().forEach(this::optimize)
```

#### Logical-to-Physical Transformation

After logical transformation, planner will let the `Table` in `LogicalRelation` (identified before logical transformation above) transform the logical plan to physical plan.

```
class OpenSearchIndex:

  def implement(plan: LogicalPlan):
    return plan.accept(
      DefaultImplementor():
        def visitNode(node):
          if node is OpenSearchLogicalIndexScan:
            return OpenSearchIndexScan(...)
          else if node is OpenSearchLogicalIndexAgg:
            return OpenSearchIndexScan(...)
```

### Problem Statement

The current planning architecture causes 2 serious problems:

1. Each data source adds special logical operator and explode the optimizer rule space. For example, Prometheus also has `PrometheusLogicalMetricAgg` and `PrometheusLogicalMetricScan` accordingly. They have the exactly same pattern to match query plan tree as OpenSearch.
2. A bigger problem is the difficulty of transforming from logical to physical when there are 2 `Table`s in query plan. Because only 1 of them has the chance to do the `implement()`. This is a blocker for supporting `INSERT ... SELECT ...` statement or JOIN query. See code below.

```
  public PhysicalPlan plan(LogicalPlan plan) {
    Table table = findTable(plan);
    if (table == null) {
      return plan.accept(new DefaultImplementor<>(), null);
    }
    return table.implement(
        table.optimize(optimize(plan)));
  }
```

#### TableScanBuilder

A new abstraction `TableScanBuilder` is added as a transition operator during logical planning and optimization. Each data source provides its implementation class by `Table` interface. The push down difference in non-aggregate and aggregate query is hidden inside specific scan builder, for example `OpenSearchIndexScanBuilder` rather than exposed to core module.

TableScanBuilder

#### TablePushDownRules

In this way, `LogicalOptimizier` in core module always have the same set of rule for all push down optimization.

LogicalPlanOptimizer

### Examples

The following diagram illustrates how `TableScanBuilder` along with `TablePushDownRule` solve the problem aforementioned.

optimizer-Page-1

Similarly, `TableWriteBuilder` will be added and work in the same way in separate PR: https://github.com/opensearch-project/sql/pull/1094

optimizer-Page-2

### TODO

1. Refactor Prometheus optimize rule and enforce table scan builder
2. Figure out how to implement AD commands
4. Deprecate `optimize()` and `implement()` if item 1 and 2 complete
5. Introduce fixed point or maximum iteration limit for iterative optimization
6. Investigate if CBO should be part of current optimizer or distributed planner in future
7. Remove `pushdownHighlight` once it's moved to OpenSearch storage
8. Move `TableScanOperator` to the new `read` package (leave it in this PR to avoid even more file changed)

### Issues Resolved
https://github.com/opensearch-project/sql/issues/948
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Yury-Fridlyand** [comment]
Is it related to #1088?

**Comment by dai-chen** [comment]
> #1088

Not related directly. I think it's related to https://github.com/opensearch-project/sql/issues/811 which is the solution?

**Comment by Yury-Fridlyand** [comment]
I like the PR description. Can you add it to the docs section?

**Comment by dai-chen** [comment]
> I like the PR description. Can you add it to the docs section?

Sure, I'm drafting another PR for updating our dev docs: https://github.com/opensearch-project/sql/pull/1092. Will add this too. Thanks!

File: core/src/main/java/org/opensearch/sql/storage/read/TableScanBuilder.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  public abstract TableScanOperator build();
+
+  /**
+   * Can a given filter operator be pushed down to table scan builder. Assume no such support
+   * by default unless subclass override this.
+   *
+   * @param filter logical filter operator
+   * @return true if pushed down, otherwise false
+   */
+  public boolean pushDownFilter(LogicalFilter filter) {
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/storage/read/TableScanBuilder.java:48]: not logical expression partial push down right?

File: core/src/main/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizer.java [hunk 2]
```diff
@@ -39,8 +41,21 @@ public LogicalPlanOptimizer(List<Rule<?>> rules) {
...
+        /*
+         * Phase 1: Transformations that rely on relational algebra equivalence
+         */
         new MergeFilterAndFilter(),
-        new PushFilterUnderSort()));
+        new PushFilterUnderSort(),
+        /*
+         * Phase 2: Transformations that rely on data source push down capability
+         */
+        new CreateTableScanBuilder(),
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizer.java:52]: Does the rule order matters?  New rules should be append after phase 2?

Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizer.java:52]: AFAIK, phase 1 has to run before phase 2. In phase 2, `CreateTableScanBuilder` rule has to precede other push down rules. However, for new rules, we may need to analyze dependency case by case.

**Review by penghuo — APPROVED** [review]

**Comment by dai-chen** [comment]
ML integration seems missing IT. Just tested manually to confirm it is not impacted.

**Review by joshuali925 — APPROVED** [review]

**Pull request merged**